jina/peapods/runtimes/grpc/__init__.py

```diff
@@ -0,0 +1,266 @@
+import time
+import traceback
+from abc import ABC
+from asyncio import CancelledError
+from collections import defaultdict
+from typing import Optional, Union
+
+from grpc import RpcError
+
+from jina.enums import OnErrorStrategy
+from jina.excepts import NoExplicitMessage, ChainedPodException, RuntimeTerminated
+from jina.helper import get_or_reuse_loop, random_identity
+from jina.peapods.grpc import Grpclet
+from jina.peapods.runtimes.base import BaseRuntime
+from jina.peapods.runtimes.request_handlers.data_request_handler import (
+    DataRequestHandler,
+)
+from jina.proto import jina_pb2
+
+from jina.types.routing.table import RoutingTable
+
+
+class GRPCDataRuntime(BaseRuntime, ABC):
+    """Runtime procedure leveraging :class:`Grpclet` for sending DataRequests"""
+
+    def __init__(self, args: 'argparse.Namespace', **kwargs):
+        """Initialize grpc and data request handling.
+        :param args: args from CLI
+        :param kwargs: extra keyword arguments
+        """
+        super().__init__(args, **kwargs)
+        self._id = random_identity()
+        self._loop = get_or_reuse_loop()
+        self._last_active_time = time.perf_counter()
+
+        self._pending_msgs = defaultdict(list)  # type: Dict[str, List['Message']]
+        self._partial_requests = None
+
+        self._data_request_handler = DataRequestHandler(args, self.logger)
+        self._grpclet = Grpclet(
+            async_loop=self._loop,
+            args=self.args,
+            message_callback=self._callback,
+            logger=self.logger,
+        )
+
+    def run_forever(self):
+        """Start the `Grpclet`."""
+        self.logger.debug('Run GRPCDataRuntime')
+        self._grpclet_task = self._loop.create_task(self._grpclet.start())
+        try:
```

Comment:
I do not think this is needed, in theory in the `__exit__` of Runtime this should be logged